src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs

```diff
@@ -186,7 +186,7 @@ public void Dispose()
             }
         }
 
-        internal object CaptureDisposable(object service)
+        internal virtual object CaptureDisposable(object service)
```

Comment:
This whole project is still pretty exotic to me; but can give it an investigation. Also I worry about the hours your team is keeping 😄 

Is `disposable == null` a common path? If so the lock portion could be factored out to a separate function which might make it more likely to inline.